runtime/major_gc.c

```diff
@@ -554,10 +563,136 @@ static void mark_ephe_aux (struct mark_stack *stk, intnat *work,
+#define Is_block_and_not_young(v) \
+  (((intnat)rotate1((uintnat)v - young_start)) > (intnat)half_young_len)
+#ifdef NO_NAKED_POINTERS
+  #define Is_major_block(v) Is_block_and_not_young(v)
+  //#define Is_major_block(v) (Is_block(v) && !Is_young(v))
+#else
+#define Is_major_block(v) (Is_block_and_not_young(v) && Is_in_heap(v))
+#endif
+
+  while (1) {
+    value *scan, *obj_end, *scan_end;
+
+    if (pb_enqueued > pb_dequeued + min_pb) {
+      /* Dequeue from prefetch buffer */
+      value block = pb[(pb_dequeued++) & Pb_mask];
+      header_t hd = Hd_val(block);
+
+      /* FIXME: Forward_tag */
+      if (Tag_hd(hd) == Forward_tag) {
+
+      } else if (Tag_hd(hd) == Infix_tag) {
+        block -= Infix_offset_val(block);
+        hd = Hd_val(block);
+      }
+
+#ifdef NO_NAKED_POINTERS
+      /* See [caml_darken] for a description of this assertion. */
+      CAMLassert (Is_in_heap (block) || Is_black_hd (hd));
+#endif
+      CAMLassert(Is_white_hd(hd) || Is_black_hd(hd));
+      if (!Is_white_hd (hd)) {
+        /* Already black, nothing to do */
+        continue;
+      }
+      // FIXME work accounting
+      hd = Blackhd_hd (hd);
+      Hd_val (block) = hd;
+      darkened_anything = 1;
+      work--; // header word
+      if (Tag_hd (hd) >= No_scan_tag) {
+        /* Nothing to scan here */
+        work -= Wosize_hd (hd);
+        continue;
+      }
+      scan = Op_val(block);
+      obj_end = scan + Wosize_hd(hd);
+
+      if (Tag_hd (hd) == Closure_tag) {
+        uintnat env_offset = Start_env_closinfo(Closinfo_val(block));
+        work -= env_offset;
+        scan += env_offset;
+      }
+    } else if (work <= 0 || stk.count == 0) {
+      if (min_pb > 0) {
+        // drain pb before quitting
+        min_pb = 0;
+        continue;
+      } else {
+        // done
+        break;
+      }
+    } else {
+      mark_entry m = stk.stack[--stk.count];
+      scan = m.start;
+      obj_end = m.end;
+    }
+
+    scan_end = obj_end;
+    work -= obj_end - scan;
+    if (work < 0) {
+      scan_end += work;
+    }
+
+    for (; scan < scan_end; scan++) {
+      value v = *scan;
+      if (Is_major_block(v)) {
+        if (pb_enqueued == pb_dequeued + Pb_size) {
+          break; /* Prefetch buffer is full */
+        }
+        caml_prefetch(Hp_val(v));
+        caml_prefetch(&Field(v, Queue_prefetch_distance - 1));
+        pb[(pb_enqueued++) & Pb_mask] = v;
+      }
+    }
+
+    if (scan < obj_end) {
+      /* Didn't finish scanning this object, either because work <= 0,
+         or the prefetch buffer filled up. Leave the rest on the stack. */
+      mark_entry m = { scan, obj_end };
+      work += obj_end - scan;
+      caml_prefetch(scan+1);
+      if (stk.count == stk.size) {
+        *Caml_state->mark_stack = stk;
+        realloc_mark_stack(Caml_state->mark_stack);
+        stk = *Caml_state->mark_stack;
+      }
+      stk.stack[stk.count++] = m;
+      min_pb = Pb_min;
```

Comment:
Why this? I would expect `min_pb` to be equal to `Pb_min` until `work` falls to (or below) 0, at which point you're going to push everything from `bp` back to `mark_stack`, so `min_pb` might as well remain 0 until you exit the loop.

Reply:
This is to do with the edge case mentioned above: min_pb becomes zero if the mark stack empties, even though that may occasionally occur when work is still large and there is lots more marking to do (the heap bottleneck case). In that case, we go into the pb-draining mode.

The point of this line is to get us out of pb-draining mode if we discover that there actually is more work to do.